vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,711 @@
+	}
+
+	connectionState, err := hostsystem.GetConnectionState(hs)
+	if err != nil {
+		return err
+	}
+
+	if connectionState == types.HostSystemConnectionStateDisconnected {
+		// Config and LicenseManager cannot be used while the host is
+		// disconnected.
+		d.Set("connected", false)
+		return nil
+	}
+	d.Set("connected", true)
+
+	lockdownMode, err := hostLockdownString(host.Config.LockdownMode)
+	if err != nil {
+		return nil
+	}
+
+	log.Printf("Setting lockdown to %s", lockdownMode)
+	d.Set("lockdown", lockdownMode)
+
+	lm := license.NewManager(client.Client)
+	am, err := lm.AssignmentManager(context.TODO())
+	if err != nil {
+		return err
+	}
+	licenses, err := am.QueryAssigned(context.TODO(), hostID)
+	if err != nil {
+		return err
+	}
+
+	licenseKey := d.Get("license").(string)
+	licFound := false
+	for _, lic := range licenses {
+		if licenseKey == lic.AssignedLicense.LicenseKey {
+			licFound = true
+			break
+		}
+	}
+
+	if !licFound {
+		d.Set("license", "")
+	}
+
+	return nil
+}
+
+func resourceVsphereHostCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+
+	hcs := buildHostConnectSpec(d)
+
+	licenseKey := d.Get("license").(string)
+
+	lm := license.NewManager(client.Client)
+	ll, err := lm.List(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	licFound := false
+	for _, l := range ll {
+		if l.LicenseKey == licenseKey {
+			licFound = true
+			break
+		}
+	}
+	if !licFound {
+		return fmt.Errorf("license key supplied (%s) did not match against known license keys", licenseKey)
+	}
+
+	var connectedState bool
+	val := d.Get("connected")
+	if val == nil {
+		connectedState = true
+	} else {
+		connectedState = val.(bool)
+	}
+
+	var task *object.Task
+	clusterID := d.Get("cluster").(string)
+	if clusterID != "" {
+		ccr, err := clustercomputeresource.FromID(client, clusterID)
+		if err != nil {
+			log.Printf("[DEBUG] Error while searching cluster %s. Error: %s", clusterID, err)
+			return err
+		}
+
+		task, err = ccr.AddHost(context.TODO(), hcs, connectedState, &licenseKey, nil)
+		if err != nil {
+			log.Printf("[DEBUG] Error while adding host with hostname %s to cluster %s.  Error: %s", d.Get("hostname").(string), clusterID, err)
+		}
+	} else {
+		dcId := d.Get("datacenter").(string)
+		dc, err := datacenterFromID(client, dcId)
+		if err != nil {
+			return err
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+		defer cancel()
+		var dcProps mo.Datacenter
+		if err := dc.Properties(ctx, dc.Reference(), nil, &dcProps); err != nil {
+			return err
+		}
+
+		hostFolder := object.NewFolder(client.Client, dcProps.HostFolder)
+		task, err = hostFolder.AddStandaloneHost(context.TODO(), hcs, connectedState, &licenseKey, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	taskResult, err := processTask(task)
+	if err != nil {
+		return fmt.Errorf("Host addition failed. %s", err)
+	}
+
+	log.Printf("[DEBUG] Task Result: %#v", taskResult)
```

Comment:
I think its generally better to avoid the fully verbose interpolation (especially with vSphere since it can get _quite_ verbose),  but there are certainly cases where it makes sense. Just want to make sure its purposeful here.

Reply:
Oops :) Yeah these are not left like that purpose. I'll do a quick sweep to be sure I haven't left any others like that.

Reply:
done